packages/babel-plugin-proposal-explicit-resource-management/src/index.ts

```diff
@@ -47,38 +47,103 @@ export default declare(api => {
       path: NodePath<t.BlockStatement | t.StaticBlock>,
       state,
     ) {
-      let stackId: t.Identifier | null = null;
-      let needsAwait = false;
-
-      for (const node of path.node.body) {
-        if (!isUsingDeclaration(node)) continue;
-        stackId ??= path.scope.generateUidIdentifier("stack");
-        const isAwaitUsing =
-          node.kind === "await using" ||
-          TOP_LEVEL_USING.get(node) === USING_KIND.AWAIT;
-        needsAwait ||= isAwaitUsing;
-
-        if (!TOP_LEVEL_USING.delete(node)) {
-          node.kind = "const";
+      if (state.availableHelper("usingCtx")) {
+        let ctx: t.Identifier | null = null;
+        let needsAwait = false;
+
+        for (const node of path.node.body) {
+          if (!isUsingDeclaration(node)) continue;
+          ctx ??= path.scope.generateUidIdentifier("usingCtx");
+          const isAwaitUsing =
+            node.kind === "await using" ||
+            TOP_LEVEL_USING.get(node) === USING_KIND.AWAIT;
+          needsAwait ||= isAwaitUsing;
+
+          if (!TOP_LEVEL_USING.delete(node)) {
+            node.kind = "const";
+          }
+          node.declarations.forEach(decl => {
+            const args = [decl.init];
+            if (isAwaitUsing) args.push(t.booleanLiteral(true));
+            decl.init = t.callExpression(
+              t.memberExpression(t.cloneNode(ctx), t.identifier("u")),
+              args,
+            );
+          });
         }
-        node.declarations.forEach(decl => {
-          const args = [t.cloneNode(stackId), decl.init];
-          if (isAwaitUsing) args.push(t.booleanLiteral(true));
-          decl.init = t.callExpression(state.addHelper("using"), args);
-        });
-      }
-      if (!stackId) return;
+        if (!ctx) return;
 
-      const errorId = path.scope.generateUidIdentifier("error");
-      const hasErrorId = path.scope.generateUidIdentifier("hasError");
+        let disposeCall: t.Expression = t.callExpression(
+          t.memberExpression(t.cloneNode(ctx), t.identifier("d")),
+          [],
+        );
+        if (needsAwait) disposeCall = t.awaitExpression(disposeCall);
 
-      let disposeCall: t.Expression = t.callExpression(
-        state.addHelper("dispose"),
-        [t.cloneNode(stackId), t.cloneNode(errorId), t.cloneNode(hasErrorId)],
-      );
-      if (needsAwait) disposeCall = t.awaitExpression(disposeCall);
+        const replacement = template.statement.ast`
+        try {
+          var ${t.cloneNode(ctx)} = ${state.addHelper("usingCtx")}();
+          ${path.node.body}
+        } catch (_) {
+          ${t.cloneNode(ctx)}.e = _;
+        } finally {
+          ${disposeCall}
+        }
+      ` as t.TryStatement;
+
+        t.inherits(replacement, path.node);
 
-      const replacement = template.statement.ast`
+        t.addComment(
+          replacement.block.body[0],
+          "leading",
+          "u: using(obj, isAwait), d: dispose()",
```

Comment:
We generally do not add comments explaining the transpiled output. Though we could do so under a `verbose: boolean` flag.

Reply:
I will remove them and open a separate PR if needed in the future.